### PR TITLE
Work around some XL issues in CUDA/OpenMP build

### PR DESCRIPTION
**Pull request by crtrott**
Errors go away by effectively chaning link order via various means (splitting CPP files, reordering how source files are added in CMake)

### CODE REVIEW

File: algorithms/unit_tests/CMakeLists.txt [hunk 1]
```diff
@@ -33,15 +33,19 @@ SET(SOURCES
...
+    TestOpenMP_SortDynamicView.cpp
+    TestOpenMP_Random.cpp
   )
 ENDIF()
 
-IF(Kokkos_ENABLE_OPENMP)
+IF(Kokkos_ENABLE_CUDA)
   LIST( APPEND SOURCES
-    TestOpenMP.cpp
+    TestCuda.cpp
```
Comment (dalg24) [inline algorithms/unit_tests/CMakeLists.txt:48]: This is messing with my head.  Why on earth is the diff displayed this way :@

File: algorithms/src/Kokkos_Sort.hpp [hunk 2]
```diff
@@ -373,15 +373,15 @@ class BinSort {
...
+  void operator()(const bin_sort_bins_tag& /*tag*/, const int i) const {
     auto bin_size = bin_count_const(i);
     if (bin_size <= 1) return;
     int upper_bound = bin_offsets(i) + bin_size;
     bool sorted     = false;
     while (!sorted) {
       sorted      = true;
       int old_idx = sort_order(bin_offsets(i));
-      int new_idx;
+      int new_idx = 0;
```
Comment (dalg24) [inline algorithms/src/Kokkos_Sort.hpp:384]: You could also have declared it inside the body of the loop.

File: algorithms/src/Kokkos_Sort.hpp [hunk 3]
```diff
@@ -349,13 +349,13 @@ class BinSort {
 
  public:
   KOKKOS_INLINE_FUNCTION
-  void operator()(const bin_count_tag& /*tag*/, const int& i) const {
```
Comment (masterleinad) [inline algorithms/src/Kokkos_Sort.hpp:352]: I guess the changes to this file are not really necessary to fix the issue?
It's probably still a good idea to pass `const` primitive types by value consistently.

Comment (crtrott) [reply algorithms/src/Kokkos_Sort.hpp:352]: essentially `i` got corrupted this was an attempt to fix it, and I thought I leave it this way.

**Pull request merged**